src/stan/io/random_var_context.hpp

```diff
@@ -0,0 +1,222 @@
+#ifndef STAN_IO_RANDOM_VAR_CONTEXT_HPP
+#define STAN_IO_RANDOM_VAR_CONTEXT_HPP
+
+#include <stan/io/var_context.hpp>
+#include <boost/random/uniform_real_distribution.hpp>
+#include <algorithm>
+#include <limits>
+#include <string>
+#include <vector>
+
+namespace stan {
+  namespace io {
+
+    /**
+     * This is an implementation of a var_context that initializes
+     * the unconstrained values randomly. This is used for initialization.
+     */
+    class random_var_context : public var_context {
+    public:
+      /**
+       * Destructor.
+       *
+       */
+      ~random_var_context() {}
+
+      /**
+       * Constructor.
+       *
+       * On construction, this var_context will generate random
+       * numbers on the unconstrained scale for the model provided.
+       * Once generated, the class is immutable.
+       *
+       * This class only generates values for the parameters in the
+       * Stan program and does not generate values for transformed parameters
+       * or generated quantities.
+       *
+       * @tparam Model a generated Stan program or a class that follows
+       *   that template
+       * @tparam RNG Random number generator type. Typically
+       *   boost::ecuyer1988
+       * @param model instantiated model to generate variables for
+       * @param[in,out] rng pseudo-random number generator
+       * @param[in] init_radius the unconstrained variables are uniform draws
+       *   from -init_radius to init_radius.
+       * @param[in] init_zero indicates whether all unconstrained variables
+       *   should be initialized at 0. When init_zero is false, init_radius
+       *   must be greater than 0.
+       */
+      template <class Model, class RNG>
+      random_var_context(Model& model,
+                         RNG& rng,
+                         double init_radius,
+                         bool init_zero)
+        : unconstrained_params_(model.num_params_r()) {
+        size_t num_unconstrained_ = model.num_params_r();
+        model.get_param_names(names_);
+        model.get_dims(dims_);
+
+        // cutting names_ and dims_ down to just the constrained parameters
+        std::vector<std::string> constrained_params_names;
+        model.constrained_param_names(constrained_params_names, false, false);
+        size_t keep = constrained_params_names.size();
+        size_t index = 0;
+        size_t num = 0;
+        for (index = 0; index < dims_.size(); index++) {
```

Comment:
yes, it's needed past the loop. I bet there's a cleaner way to write the function.